index.js

```diff
@@ -0,0 +1,84 @@
+
+require('dotenv').config();
+const bcrypt = require('bcryptjs');
+const express = require('express');
+const helmet = require('helmet');
+
+const server = express();
+server.use(helmet());
+server.use(express.json());
+
+const restrictedRouter = require('./restricted-router.js');
+server.use('/api/restricted', authCheck, restrictedRouter);
+
+const db = require('./data/dbConfig.js');
+
+server.get('/api/users', authCheck, async (req,res) => {
+    try {
+        const result = await db('users');
+        res.status(200).json(result);
+    }
+    catch(err) {
+        res.status(500).json({err});
+    }
+});
+
+server.post('/api/register', async (req,res) => {
+    let user = req.body;
+    if (!user.username || !user.password) {
+        res.status(400).json({error:'please provide username/password'});
+    } else {
+        user.password = bcrypt.hashSync(user.password, 8);
+        try {
+            const result = await db('users').insert(user);
+            if (result) {
+                const newUser = await db('users').where({'id':result[0]}).first();
+                res.status(201).json(newUser);
+            }
+        }
+        catch(err) {
+            res.status(500).json(err);
+        }
+    }
+});
+
+server.post('/api/login', async (req,res) => {
+    let {username,password} = req.body;
+    if (!username || !password) {
+        res.status(400).json({error:'please provide username/password'});
+    } else {
+        try {
+            const validUser = await db('users').where({username}).first();
+            if (validUser && bcrypt.compareSync(password, validUser.password)) {
+                res.status(200).json({message:`${validUser.username} LOGGED IN`});
+            } else {
+                res.status(401).json({message:'invalid credentials'});
+            }
+        }
+        catch(err) {
+            res.status(500).json(err);
+        }
+    }
+});
+
+async function authCheck(req,res,next) {
+    const {username,password} = req.headers;
+
+    try {
+        const validUser = await db('users').where({username}).first();
+        if (validUser && bcrypt.compareSync(password, validUser.password)) {
+            next();
+        } else {
+            res.status(401).json({message:'invalid credentials'});
+        }
```

Comment:
yes, perfect implementation of the restricted middleware! remember why it's important that you return 401 if either falsey case in your conditional happens, but that you DON'T tell the user which case happened -- you are the gatekeeper, and your job here is to supply as little info as possible to those who may have bad intentions with your database